tests/e2e/conftest.py

```diff
@@ -109,3 +122,102 @@ def _f(cluster: str) -> Iterator[None]:
                 )
 
     return _f
+
+
+@pytest.fixture
+def dockerhub_auth_secret(_neuro_client: neuro_api.Client) -> Iterator[Secret]:
+    secret_name = f"{KANIKO_AUTH_PREFIX}_{uuid.uuid4().hex}"
+    auth_data = _construct_registy_auth(
+        # Why not v2: https://github.com/GoogleContainerTools/kaniko/pull/1209
+        registry_uri="https://index.docker.io/v1/",
+        username=os.environ["DOCKER_CI_USERNAME"],
+        password=os.environ["DOCKER_CI_TOKEN"],
+    )
+    secret = Secret(secret_name, auth_data)
+    try:
+        run_async(_neuro_client.secrets.add(secret_name, auth_data.encode()))
+        yield secret
+    finally:
+        run_async(_neuro_client.secrets.rm(secret_name))
+
+
+@pytest.fixture
+def project_dir() -> Iterator[Path]:
+    with TemporaryDirectory() as cwd_str:
+        old_cwd = Path.cwd()
+        cwd = Path(cwd_str)
+        os.chdir(cwd)
+        try:
+            yield cwd
+        finally:
+            os.chdir(old_cwd)
+
+
+@pytest.fixture()
+def cli_runner(capfd: CaptureFixture[str], project_dir: Path) -> CLIRunner:
+    def _run_cli(args: List[str]) -> "CompletedProcess[str]":
+        args = args.copy()
+        cmd = args.pop(0)
+        if cmd not in ("neuro", "neuro-extras"):
+            pytest.fail(f"Illegal command: {cmd}")
+
+        run_cmd = f"Run '{cmd} {' '.join(args)}'"
+        logger.info(run_cmd)
+        capfd.readouterr()
+
+        main = extras_main
+        if cmd == "neuro":
+            args = [
+                "--show-traceback",
+                "--disable-pypi-version-check",
+                "--color=no",
+            ] + args
+            main = neuro_main
+
+        code = EX_OK
+        try:
+            main(args)
+        except SystemExit as e:
+            code = e.code
+        out, err = capfd.readouterr()
+        out, err = out.strip(), err.strip()
+        if code != EX_OK:
+            # This generates too much garbage, but might be handy for debugging
+            logger.info(f"Stdout:\n{SEP_BEGIN}\n{out}\n{SEP_END}\nStdout finished")
+        if err:
+            logger.info(f"Stderr:\n{SEP_BEGIN}\n{err}\n{SEP_END}\nStderr finished")
+
+        return CompletedProcess(
+            args=[cmd] + args, returncode=code, stdout=out, stderr=err
+        )
+
+    return _run_cli
+
+
+@pytest.fixture
+def repeat_until_success(
```

Comment:
Consider using `tenacity` for timed retries until success: https://tenacity.readthedocs.io/en/latest/

If it's difficult to integrate here, feel free to ignore

Reply:
great idea, added in a0dbccb4ab